File: libs/loomj/src/main/java/dev/promptloom/loomj/Safe.java
package dev.promptloom.loomj;

import java.util.logging.Logger;

/**
 * Fluent API for safely loading secrets in LoomLocker-protected applications.
 *
 * <p>Usage:
 * <pre>{@code
 * new Safe()
 *     .unlock()
 *     .execute(() -> loadEnv())
 *     .autolock();
 *
 * server.start();
 * }</pre>
 *
 * <p>If loomlocker is not running, all operations are transparent no-ops —
 * {@link #execute(Runnable)} always runs.
 *
 * <p>Password resolution order: {@link #unlock(String)} argument →
 * {@code LOOM_SESSION_PASSWORD} env var → skip (secrets stay locked).
 */
public class Safe {

    private static final Logger log = Logger.getLogger(Safe.class.getName());

    private final LockerConfig config;
    private final LoomLockerClient client;
    private boolean unlocked = false;
    private boolean silent = false;

    /** Create a Safe with config auto-detected from {@code .loom.config} or env vars. */
    public Safe() {
        this.config = LockerConfig.fromEnv();
        this.client = new LoomLockerClient(this.config);
    }

    /** Create a Safe with an explicit config. */
    public Safe(LockerConfig config) {
        this.config = config;
        this.client = new LoomLockerClient(config);
    }

    /** Suppress all log output from this Safe instance. */
    public Safe silent() {
        this.silent = true;
        return this;
    }

    // ── Fluent API ────────────────────────────────────────────────────────────

    /**
     * Unlock secrets using a password.
     * Falls back to {@code LOOM_SESSION_PASSWORD} env var if no argument supplied.
     * No-op if loomlocker is not running or secrets are already unlocked.
     */
    public Safe unlock(String... password) {
        if (!client.isRunning()) {
            return this;
        }
        if (!client.isLocked()) {
            unlocked = true;
            return this;
        }
        String pwd = (password.length > 0 && password[0] != null) ? password[0] : "";
        if (pwd.isEmpty()) {
            pwd = System.getenv("LOOM_SESSION_PASSWORD");
            if (pwd == null) pwd = "";
        }
        if (pwd.isEmpty()) {
            info("secrets are locked but no password provided — set LOOM_SESSION_PASSWORD");
            return this;
        }
        if (client.unlock(pwd)) {
            unlocked = true;
        } else {
            info("unlock failed: wrong password or server error");
        }
        return this;
    }

    /**
     * Run {@code fn}. Always called regardless of lock state — secrets loading
     * (e.g. {@code dotenv}) happens inside fn.
     */
    public Safe execute(Runnable fn) {
        fn.run();
        return this;
    }

    /**
     * Signal that startup is complete, triggering an immediate relock on the server.
     * No-op if loomlocker is not running or was never unlocked.
     */
    public Safe autolock() {
        if (unlocked && client.isRunning()) {
            if (!client.autolock()) {
                info("autolock request failed");
            }
            unlocked = false;
        }
        return this;
    }

    // ── Internal ──────────────────────────────────────────────────────────────

    private void info(String msg) {
        if (!silent) {
            log.info("[loomj] " + msg);
        }
    }
}

File: libs/bloompy/bloompy/safe.py
"""The Safe class — the primary user-facing API for bloompy."""
from __future__ import annotations

import logging
import os
from typing import Callable, TypeVar

from .client import LoomLockerClient
from .config import LockerConfig

logger = logging.getLogger(__name__)
T = TypeVar("T")


class Safe:
    """Fluent API for safely loading secrets in loomlocker-protected projects.

    Basic usage::

        from bloompy import Safe

        # Simple: unlock → run → autolock
        Safe().unlock().execute(lambda: load_dotenv()).autolock()

        # Explicit password (falls back to LOOM_SESSION_PASSWORD env var)
        Safe().unlock("my-password").execute(load_dotenv).autolock()

    If loomlocker is not running, all operations are transparent no-ops —
    the ``execute`` function always runs.
    """

    def __init__(
        self,
        config: LockerConfig | None = None,
        client: LoomLockerClient | None = None,
    ) -> None:
        self._cfg = config or LockerConfig.from_env()
        self._client = client or LoomLockerClient(self._cfg)
        self._unlocked = False
        self._silent = False

    # ── Fluent API ────────────────────────────────────────────────────────────

    def silent(self) -> "Safe":
        """Suppress all warnings from this Safe (debug logging is unaffected)."""
        self._silent = True
        return self

    def _warn(self, message: str, *args) -> None:
        if not self._silent:
            logger.warning(message, *args)

    def unlock(self, password: str | None = None) -> "Safe":
        """Unlock secrets if loomlocker is running and they are locked.

        Password precedence:
        1. ``password`` argument
        2. ``LOOM_SESSION_PASSWORD`` environment variable
        3. No-op (server not running or already unlocked)
        """
        if not self._client.is_running():
            logger.debug("bloompy: loomlocker not running — skipping unlock")
            return self

        if not self._client.is_locked():
            logger.debug("bloompy: secrets already unlocked")
            self._unlocked = True
            return self

        pwd = password or os.getenv("LOOM_SESSION_PASSWORD", "")
        if not pwd:
            self._warn(
                "bloompy: secrets are locked but no password provided. "
                "Set LOOM_SESSION_PASSWORD or pass password= to unlock()."
            )
            return self

        if self._client.unlock(pwd):
            logger.debug("bloompy: secrets unlocked (auto-relock in %ds)",
                         self._cfg.unlock_duration_seconds)
            self._unlocked = True
        else:
            self._warn("bloompy: unlock failed — check password")

        return self

    def execute(self, fn: Callable[[], T]) -> "Safe":
        """Run *fn* and return self for chaining. Always called regardless of lock state."""
        fn()
        return self

    def autolock(self) -> "Safe":
        """Signal loomlocker that startup is complete → trigger immediate relock.

        If loomlocker is not running or unlock() was not called, this is a no-op.
        """
        if self._unlocked and self._client.is_running():
            self._client.autolock()
            self._unlocked = False
            logger.debug("bloompy: autolock requested")
        return self

    # ── Context manager support ───────────────────────────────────────────────

    def __enter__(self) -> "Safe":
        return self.unlock()

    def __exit__(self, *_) -> None:
        self.autolock()
